client/src/components/DropdownMenu.tsx

```diff
@@ -1,22 +1,44 @@
 import classNames from "classnames";
 import _ from "lodash";
-import React from "react";
+import React, { BaseSyntheticEvent } from "react";
 
-import "./DropdownMenu.scss";
 import { trivial_text_maker } from "src/models/text.js";
 
-export class DropdownMenu extends React.Component {
-  constructor(props) {
+import "./DropdownMenu.scss";
+
+interface DropdownMenuProps {
+  dropdown_content: React.ReactNode;
+  opened_button_class_name: string;
+  closed_button_class_name: string;
+  dropdown_trigger_txt: React.ReactNode | string;
+  dropdown_content_class_name?: string;
+  button_description?: string;
+  dropdown_a11y_txt?: string;
+}
+
+type DropdownMenuState = {
+  is_open: boolean;
+};
+
+export class DropdownMenu extends React.Component<
+  DropdownMenuProps,
+  DropdownMenuState
+> {
+  dropdown_ref = React.createRef<HTMLDivElement>();
+  toggle_dropdown_button = React.createRef<HTMLDivElement>();
+  dropdown_area = React.createRef<HTMLDivElement>();
+
+  constructor(props: DropdownMenuProps) {
     super(props);
     this.state = {
       is_open: false,
     };
-
-    this.dropdown_ref = React.createRef();
   }
-  handleWindowClick = (e) => {
+
+  handleWindowClick = (e: MouseEvent) => {
     const { is_open } = this.state;
-    if (is_open && !this.dropdown_ref.current.contains(e.target)) {
+    const dropdown_node = this.dropdown_ref.current;
+    if (is_open && dropdown_node && !dropdown_node.contains(e.target as Node)) {
```

Comment:
https://github.com/Microsoft/TypeScript/issues/15394#issuecomment-297495746
```
Basically EventTarget is the most general type, of which Element is a subtype, and HTMLElement is a subtype of that. If you get back a thing from the DOM, we generally have no idea which it is, and you should add a type assertion to "add in" the specific external knowledge that you have about the structure of your particular DOM layout.

It's possible, for example, that the .target of an event is not an Element (you can add event listeners to XMLHttpRequest, for example, and XMLHttpRequest does not have a getBoundingClientRect method).
```
Basically: target may or may not be a Node, so TS can't infer. Hence, need to assert `as Node`